### DISCUSSION TITLE
Graceful shutdown

### DISCUSSION THREAD

**Discussion by nkocyigit** [jobrunr/jobrunr#929] [Ideas]
Hello,
It would be nice to have a feature like graceful shutdown, in case a background job server is initiated to shut down it may wait until it finishes all the jobs it is currently processing. This could be a boolean property that can be set. It is a useful feature in environments where background job servers are killed e.g., in a cluster environment.
thanks

**Comment by rdehuyss** [comment]
This is possible in the Pro version.

**Reply by yuanwang-wf** [reply]
hey, want to follow up on this subject. is there any documentation or code snippet on how to enable in pro version? thank you!

**Reply by auloin** [reply]
Hi @yuanwang-wf, you can use the property `background-job-server.interrupt-jobs-await-duration-on-stop`.

E.g. for Spring Boot,`org.jobrunr.background-job-server.interrupt-jobs-await-duration-on-stop=PT25S`. This example, tells the `BackgroundJobServer` to wait about 25 seconds for the jobs to finish (will exit earlier if the job completes faster).

**Reply by yuanwang-wf** [reply]
hey @auloin thank you for quick response. i see. it does shutdown ` BackgroundJobPerformers` gracefully. For zookeeper thread's tasks, it seems the shutdown on that thread group is hardcore to 10second. we have custom logic for recurring jobs. During deploy, we see logs like

```
Recurring job 'JobName' resulted in 1 scheduled jobs in time range 2024-11-22T16:37:00Z - 2024-11-22T16:37:00Z (PT0S) but it is already SCHEDULED, AWAITING, ENQUEUED or PROCESSING. Run will be skipped as job is taking longer than given schedule (CronExpression, Interval or CustomSchedule). 
```

i wonder is there some guideline on graceful shutdown the zookeeper tasks. thank you. please let me if this is not the best channel for this question. once again, thank you.

**Reply by auloin** [reply]
@yuanwang-wf can you explain why this is an issue for you?

**Reply by yuanwang-wf** [reply]
within a deploy cycle, we observed some recurring tasks missed one round, if that makes sense. I think the missing gap is caused by old version's master node shutdown and its zookeeper task, and new version's master node start to run

**Reply by auloin** [reply]
@yuanwang-wf  You're probably right about the missing gap being caused by the shutdown. But I don't see how graceful shutdown of the zookeepers would fix it (see my reasoning below). Is there something with your use case that makes a graceful shutdown of zookeepers desirable?

Graceful shutdown here means that the workers will no longer fetch and process new work, they will finish those already processing and others will remain in the queue. So any new Job created by a RecurringJob, right before or during shutdown, will remain in the queue (`SCHEDULED` or `ENQUEUED`). 

Also the zookeepers are either creating new jobs (from the recurring jobs) or changing the states of existing jobs, so no processing. This means they can stop as soon as possible (10 seconds of shutdown period is almost always enough for them to finish whatever work they were doing). From this perspective, I don't see what they should be waiting for...

For most use cases, the log you're seeing on restart is ok as it means the RecurringJob already has an instance in the queue, processing or waiting to be picked up (so no need to create a new one).

There is the `maxConcurrentJobs` attribute that can allow for more than one instance of a RecurringJob. Is this something that would work for you? Or are you looking for a different graceful mechanism? If so it would be great to have a detailed explanation.

**Reply by yuanwang-wf** [reply]
thank you @auloin . it seems `maxConcurrentJobs` might be the way to solve my issue.
